launcher/src/main/java/org/apache/spark/launcher/AbstractLauncher.java

```diff
@@ -0,0 +1,600 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.launcher;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileFilter;
+import java.io.FileInputStream;
+import java.io.InputStreamReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.regex.Pattern;
+
+/**
+ * Basic functionality for launchers.
+ */
+public abstract class AbstractLauncher<T extends AbstractLauncher> extends LauncherCommon {
+
+  protected static final String DEFAULT_MEM = "512m";
+
+  protected String javaHome;
+  protected String sparkHome;
+  protected String propertiesFile;
+  protected final Map<String, String> conf = new HashMap<String, String>();
+  private final Map<String, String> env;
+
+  protected AbstractLauncher() {
+    this(null);
+  }
+
+  protected AbstractLauncher(Map<String, String> env) {
+    this.env = env;
+  }
+
+  @SuppressWarnings("unchecked")
+  private final T THIS = (T) this;
+
+  /** Set a custom JAVA_HOME for launching the Spark application. */
+  public T setJavaHome(String path) {
+    checkNotNull(path, "path");
+    this.javaHome = path;
+    return THIS;
+  }
+
+  /** Set a custom Spark installation location for the application. */
+  public T setSparkHome(String path) {
+    checkNotNull(path, "path");
+    this.sparkHome = path;
+    return THIS;
+  }
+
+  /** Set a custom properties file with Spark configuration for the application. */
+  public T setPropertiesFile(String path) {
+    checkNotNull(path, "path");
+    this.propertiesFile = path;
+    return THIS;
+  }
+
+  /** Set a single configuration value for the application. */
+  public T setConf(String key, String value) {
+    checkNotNull(key, "key");
+    checkNotNull(value, "value");
+    checkArgument(key.startsWith("spark."), "'key' must start with 'spark.'");
+    conf.put(key, value);
+    return THIS;
+  }
+
+  /**
+   * Launchers should implement this to create the command to be executed.
+   */
+  protected abstract List<String> buildLauncherCommand() throws IOException;
+
+  protected Properties loadPropertiesFile() throws IOException {
+    Properties props = new Properties();
+    File propsFile;
+    if (propertiesFile != null) {
+      propsFile = new File(propertiesFile);
+      checkArgument(propsFile.isFile(), "Invalid properties file '%s'.", propertiesFile);
+    } else {
+      String confDir = getenv("SPARK_CONF_DIR");
+      if (confDir == null) {
+        confDir = join(File.separator, getSparkHome(), "conf");
+      }
+      propsFile = new File(confDir, "spark-defaults.conf");
```

Comment:
could you make `spark-defaults.conf` a static variable `DEFAULT_PROPERTIES_FILE` or something